Custom grafana dashboard

jbasu01 opened the pull request:
A new custom Grafana dashboard was created and tested by bringing in metrics from ArgoCD. ArgoCD provides a JSON template download, and code from the JSON download could be copied into the config file and appropriate changes made to ensure metrics from ArgoCD dashboard are pulled in and made visible in the custom Grafana dashboard. Additional information is available at https://grafana.com/grafana/dashboards/14584-argocd/

A key item to make a note of is that one cannot use the word "ArgoCD" or "argocd" just by itself in the value for the key "title" in the grafana-dashboard-argocd.yaml.  When changed to "ArgoCD xxxxx" ("ArgoCD stats" in our case and it worked fine.

jbasu01 committed:
Create a working custom-grafana-dashboard for ArgoCD metrics

This will allow cluster-admins, nerc-org-admins, and nerc-ops teams to
develop new dashboards using the existing multi-cluster observability
metrics.

rh-pre-commit.version: 2.0.3
rh-pre-commit.check-secrets: ENABLED

schwesig reviewed (approved):
working together with Jeet in a call
checking and linting together

computate commented on grafana/base/grafanadatasources/grafanadatasource.yaml:
```diff
@@ -0,0 +1,38 @@
+spec:
+  name: loki
+  datasources:
+    - name: application-logging
```
We need to remove the Loki datasources and replace it with the prometheus datasource instead, and make sure the Patch Operator is working to patch the Bearer token in here: 
```yaml
  datasources:
    - access: proxy
      editable: false
      isDefault: true
      jsonData:
        httpHeaderName1: Authorization
        timeInterval: 5s
        tlsSkipVerify: true
      name: Prometheus
      secureJsonData:
        httpHeaderValue1: >-
          Bearer ...
      type: prometheus
      url: 'https://thanos-querier.openshift-monitoring.svc.cluster.local:9091/'
  name: custom-observability-grafana
  ```

larsks commented on grafana/base/grafanadatasources/grafanadatasource.yaml:
```diff
@@ -0,0 +1,38 @@
+        https://lokistack-querier-http.openshift-operators-redhat.svc.cluster.local:3100
+      withCredentials: true
+      jsonData:
+        tlsSkipVerify: true
```
Are you sure this is necessary? You should be able to use the ca cert provided in `/run/secrets/kubernetes.io/serviceaccount` to trust other services on the cluster without disabling tls verification.

larsks commented on grafana/base/grafanas/grafana.yaml:
```diff
@@ -0,0 +1,26 @@
+          name: grafana-config-overrides
+  config:
+    server:
+      root_url: $GRAFANA_ROOT_URL
```
I would avoid placeholders that look like variables (because some might be confused about how things operate). Consider something like...

```suggestion
      root_url: REPLACE_IN_OVERLAY
```

computate replied:
I have applied this suggestion.

larsks commented on grafana/base/kustomization.yaml:
```diff
@@ -0,0 +1,14 @@
+  - subscriptions
+  - configmaps
+  - grafanas
+  #  - grafanadatasources
```
If we don't need this, remove it. If it's meant as an example, then add a comment documenting why it's here.

computate replied:
I have applied this suggestion.

larsks commented on grafana/base/routes/route.yaml:
```diff
@@ -0,0 +1,17 @@
+  name: grafana
+  namespace: grafana
+spec:
+  host: grafana.apps.nerc-ocp-infra.rc.fas.harvard.edu
```
You should treat this the same way you treat `root_url` in `base/grafanas/grafana.yaml`.

computate replied:
I have applied this suggestion.

larsks commented on grafana/base/serviceaccounts/serviceaccount.yaml:
```diff
@@ -0,0 +1,6 @@
+---
```
Remove the YAML document separator (you're not using it anywhere else).

computate replied:
I have applied this suggestion.

larsks commented on grafana/overlays/nerc-ocp-infra/kustomization.yaml:
```diff
@@ -0,0 +1,10 @@
+  - patches/logging-grafana-patch.yaml
+
+patchesStrategicMerge:
+  - grafanas/logging-grafana_patch.yaml
```
This should be:

```suggestion
patches:
- path: grafanas/logging-grafana_patch.yaml
```

`patchesStrategicMerge` has been [deprecated for a while](https://github.com/kubernetes-sigs/kustomize/issues/5052).

larsks committed:
Add missing redirect URL for grafana

We're reusing the dex configuration for logging-grafana, but we had not
updated the list of valid redirect urls in Dex, so oauth logins were
failing.

larsks committed:
Replace static service account token

Previously we were storing a service account token in the vault and
retrieving it via an ExternalSecret. This is not necessary; Kubernetes
already has the ability to populate a Secret with the token for a service
account [1].

[1]: https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#create-token

larsks committed:
Remove live patches

This commit removes the patch-operator [1] based patches and accompanying
support resources such as service accounts, RBAC, etc.

[1]: https://github.com/redhat-cop/patch-operator

larsks committed:
Removed unused ConfigMap

We're not using the grafana-config-overrides ConfigMap.

larsks committed:
Rename secret key containing oauth client secret

Grafana supports setting configuration from environment variables [1]. This
commit renames the secret key in the oauth-client-secret resource to match
the required variable name pattern so that Grafana will get the oauth
client secret from the environment.

[1]: https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#override-configuration-with-environment-variables

larsks committed:
Rename and reconfigure the observability-metrics datasource

Grafana supports environment variable substitution in data sources [1].
This commit modifies the observability-metrics datasource to get the
service account token and CA certificate from the environment, rather than
using a live patch.

[1]: https://grafana.com/docs/grafana/latest/administration/provisioning/#using-environment-variables

larsks committed:
Configure grafana to read environment variables from secrets

This commit modifies the Grafana resource to read environment variables
from the oauth-client-secret and grafana-serviceaccount-token Secrets and
the openshift-service-ca.crt ConfigMap.

larsks committed:
Set static Grafana configuration

The server root_url and the oauth client id are not dynamic nor are they
secret, so we can set them statically rather than patching them.

larsks committed:
Make some minor formatting changes

This commit has some minor formatting changes that didn't
really fit in anywhere else.

computate reviewed (approved):
@larsks Nice work with the secretGenerator, ExternalSecrets, and environment variable substitutions, thanks!

schwesig reviewed (approved):
@larsks awesome work, thanks for having these ideas/digging them up. appreciated.
Also thanks for getting me all the details in our call.

computate commented on grafana/overlays/nerc-ocp-infra/grafanadatasources/observability-metrics.yaml:
```diff
@@ -0,0 +1,21 @@
+        httpHeaderValue1: "Bearer ${token}"
+        tlsCACert: "${service-ca.crt}"
+      type: prometheus
+      url: 'https://thanos-querier.openshift-monitoring.svc.cluster.local:9091/'
```
@harshil-codes pointed out that we should change the url here to be the correct observability thanos querier `http://observability-thanos-query.open-cluster-management-observability.svc.cluster.local:9090`

computate replied:
@schwesig @larsks  I just added a commit to change this URL while making another change to add a ClusterRoleBinding for the grafana-servicaccount to read the metrics.

computate committed:
ClusterRoleBinding for grafana-serviceaccount to cluster-monitoring-view

computate committed:
Point GrafanaDataSource to correctly point to Observability
